src/abi.md

```diff
@@ -1,21 +1,246 @@
+> ABI compatibility is a pairwise relation between two types. With Transivity and Symmetry, however, it can be well-defined to refer to several types being ABI compatible
+
+r[abi.compatibility.integer]
+Two [integer types] are *abi compatible* if they have the same width and the same signedness.
+
+> [!NOTE]
+> The width of an integer type is the number of bits, e.g. `u8` has a width of 8, and `i128` has a width of 128.
+
+> [!NOTE]
+> In particular, [`usize`] is *abi compatible* with `uN`, and `isize` is *abi compatible* with `iN` where `N` is the target_pointer_width.
+> Two integer types with different signedness, such as `u8` and `i8` are not *abi compatible*.
+
+```rust
+#[cfg(target_pointer_width = "32")]
+fn foo(x: u32) -> u32 {
+    x
+}
+#[cfg(target_pointer_width = "64")]
+fn foo(x: u64) -> u64 {
+    x
+}
+
+fn main() {
+    let f: fn(usize) -> usize = unsafe { core::mem::transmute(foo as fn(_) -> _) };
+    let x = 0usize;
+    let y = f(x);
+    assert_eq!(x, y);
+}
+```
+
+r[abi.compatibility.char]
+The type [`char`] is *abi compatible* with the type [`u32`][integer types].
+
+```rust
+fn foo(x: char) -> u32{
+    x as u32
+}
+
+fn main(){
+    let f: fn(u32)->char = unsafe{core::mem::transmute(foo as fn(_)->_)};
+    let x = b'A' as u32; // ascii character indices are the same as Unicode character indices
+    let y = f(x);
+    assert_eq!(y, 'A');
+}
+```
+
+r[abi.compatibility.pointer]
+Two [pointer types], `*mut T` and `*const U`, are *abi compatible* if the *metadata type*s of `T` and `U` are the same type.
+
+> [!NOTE]
+> [`core::marker::Sized`] types have a *metadata type* of `()`.
+
+> [!NOTE]
+> With transitivity, this applies regardless of the mutability of either pointer type
+
+```rust
+unsafe fn foo(x: *mut u32){
+   unsafe{x.write(5);}
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function
+    let mut val: u32 = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.reference-box]
+The types [`&T`], [`&mut T`], [`alloc::boxed::Box<T>`], and [`core::ptr::NonNull<T>`], are *abi compatible* with `*const T`
+
+> [!NOTE]
+> With transitivity, they are also *abi compatible* with each other, and with `*mut T`, as well as references/`Box` to different types that have the same *metadata type*.
+
+```rust
+fn foo(x: &mut i32){
+   *x = 5;
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as fn(_))}; // Type Erase the function
+    let mut val = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.core]
+The types [`core::mem::MaybeUninit<T>`], [`core::cell::UnsafeCell<T>`], and [`core::num::NonZero<T>`], are *abi compatible* with `T`.
```

Comment:
This is odd... why do we list these types and not others?

Reply:
These types are magic lang items. `MaybeUninit<T>` is a transparent union (which isn't otherwise a thing that can exist), `UnsafeCell<T>` doesn't fully qualify as normal `repr(transparent)` due to excluding niches (including guaranteed ones that would normally be inherited by `repr(transparent)`, and `NonZero<T>` has GDE.